{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9f28f7ee",
   "metadata": {},
   "source": [
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d007401",
   "metadata": {},
   "source": [
    "# <center>Notebook to launch **`Stat-SDS`** analysis </center>\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "    \n",
    "<b>Note:</b> Part of the **`Stat-SDS`** workflow below is adapted for running on a remote server with enough computing power. If running locally, you will have to adapt the procedure (and the scripts) to your architecture.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbba5f0c",
   "metadata": {},
   "source": [
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44e2ea8b",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    \n",
    "<center><b> Warning:</b> Before running this notebook you need to set up both the server and local directories you will be working with. </center></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a552edba",
   "metadata": {},
   "source": [
    "## Server setup\n",
    "Create a folder named **`StatSDS`**. \n",
    "\n",
    "Copy **ALL THE PYTHON SCRIPTS** contained in the **`StatSDS_SCRIPTS`** folder of the present workflow to the **`StatSDS`** folder created in the server.  \n",
    "\n",
    "Copy the **`mousenet_V2_filtered`** folder (the folder **AND** its contents, **NOT ONLY** the contents) to the **`StatSDS`** folder created in the server.\n",
    "\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The structure in the server has to be:</b>  \n",
    "    \n",
    "-  **`StatSDS`**\n",
    " - **`mousenet_V2_filtered`**\n",
    "     - `geneNames.txt`\n",
    "     - `networkNames.txt`\n",
    " - **`fourier.py`**\n",
    " - **`genextract.py`**\n",
    " - **`launch_statSDS_V1.py`**\n",
    " - **`parameters.py`**\n",
    " - **`sds.py`**\n",
    " - **`sparse.py`**\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f01f227",
   "metadata": {},
   "source": [
    "## Local setup\n",
    "\n",
    "Create a folder for future storage of **`StatSDS`** analysis results in your local computer:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa80870d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "outdir = './ANALYSIS_RESULTS/'\n",
    "\n",
    "CHECK_FOLDER = os.path.isdir(outdir)\n",
    "if not CHECK_FOLDER:\n",
    "    os.makedirs(outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58a8e8f7",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The script creates:</b>   \n",
    "    \n",
    "A folder called **`ANALYSIS_RESULTS`** in the **`statSDS`** directory where this notebook is located.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Thus, the structure in the local computer at this point of the analysis is:</b>  \n",
    "    \n",
    "-  **`statSDS`**\n",
    " - **`ANALYSIS_RESULTS`** \n",
    " - **`EXAMPLE_INPUT_DATA`**\n",
    " - **`mousenet_V2_filtered`**\n",
    " -  **`statSDS_SCRIPTS`**     \n",
    " - **`SCRIPTS`** \n",
    " - **`launch_staSDS.ipynb`** \n",
    "    \n",
    "</div>\n",
    "    </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1578f825",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa8168b2",
   "metadata": {},
   "source": [
    "## 1 - Generate bootstrap replicates of the experiment in local computer\n",
    "\n",
    "In order to generate bootstrap replicates of the experiment, you need to know the number of experimental replicates in your **`CSV`** count data file that describes the experiment and parse it under the  **`--n_experimental_replicates`** argument of the script.  \n",
    "\n",
    "You will also need to pass the number of bootstrap replicates you want to calculate under the \n",
    "**`--n_bs_replicates`** argument of the script."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "590b9a2a",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "    \n",
    "Indeed, your data file should be structured as the provided example data file **`EXAMPLE_INPUT_DATA/example_read_count_file.csv`** , where the replicate number is enclosed in the column **` {DISEASECONDITION}_{REPLICATENUMBER}_counts `**.   \n",
    "\n",
    "The example data file contains data for 10 replicates (numbered from 1 to 10) of **`RNAseq`** data counts of a single cell type (Drd1 neurons) for 5 different disease conditions, namely the polyQ length of Huntington disease mutations Q20 50, 111, 170 and 175 ([Lee et al. 2020](https://pubmed.ncbi.nlm.nih.gov/32681824/)). Q20 is used further down the analysis as control for LFC value calculation (see `R` scripts).</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b9bec72",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "<b>Warning:</b> The script is by default run for the example count data file provided with this notebook. \n",
    "To adapt it to another input count file, you will need to change the line of code 105 of the script.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6f88ca93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of experimental replicates = 10\n",
      "Number of bootstrap replicates of the experiment to generate = 3\n",
      " ---------------------------------------\n",
      "    Generating BS replicate number  1\n",
      " ---------------------------------------\n",
      " ---------------------------------------\n",
      "    Generating BS replicate number  2\n",
      " ---------------------------------------\n",
      " ---------------------------------------\n",
      "    Generating BS replicate number  3\n",
      " ---------------------------------------\n"
     ]
    }
   ],
   "source": [
    "%run ./SCRIPTS/1_bootstrap_mouse_replicates_Jupyter --n_experimental_replicates 10 --n_bs_replicates 3 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1205744",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The script creates:</b> \n",
    "    \n",
    "A folder named **`BOOTSTRAP_REPLICATES_OF_EXPERIMENT`** within the **`ANALYSIS_RESULTS`** folder previously created, where the amount of bootstrap replicates **`--n_bs_replicates`** chosen by the user are generated from the number of replicates in the experimental data **`--n_experimental_replicates`**.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Thus, the structure in the local computer at this point of the analysis is:</b>  \n",
    "    \n",
    "-  **`statSDS`**\n",
    " - **`ANALYSIS_RESULTS`** \n",
    "     - **`BOOTSTRAP_REPLICATES_OF_EXPERIMENT`**\n",
    "        - `bootstrap_replicates_ids.txt`\n",
    "        - `BSreplicate_1.csv`\n",
    "        - `BSreplicate_2.csv`\n",
    "        - `BSreplicate_3.csv`\n",
    "    \n",
    " - **`EXAMPLE_INPUT_DATA`**\n",
    " - **`mousenet_V2_filtered`**\n",
    " -  **`statSDS_SCRIPTS`**     \n",
    " - **`SCRIPTS`** \n",
    " - **`launch_staSDS.ipynb`** \n",
    "    \n",
    "</div>\n",
    "    </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b9e71d0",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4edee1c2",
   "metadata": {},
   "source": [
    "##  2 - Run `DESeq2` in local computer on count data of the bootstrap replicates of the experiment and on count data of original experiment "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f5aa723",
   "metadata": {},
   "source": [
    "Run **`/SCRIPTS/2_get_metadata_BSreplicates_of_experiment_Jupyter.r`** in **`Rstudio`** or using **`R`** in your terminal.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The script creates:</b>\n",
    "    \n",
    "- A folder called **`PRE_PROCESSING_BS_REPLICATES`** in the root directory.\n",
    "- Several files necessary for **`DESEq2`** analyses inside **`PRE_PROCESSING_BS_REPLICATES`**</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5184207b",
   "metadata": {},
   "source": [
    "Run **`/SCRIPTS/2_get_metadata_original_experiment_Jupyter.r`** in **`Rstudio`** or using  **`R`** in your terminal.\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The script creates:</b><br>   \n",
    "    \n",
    "- A folder called **`PRE_PROCESSING_ORIGINAL_EXPERIMENT`** in the root directory<br>\n",
    "- Several files necessary for **`DESEq2`** analyses inside **`PRE_PROCESSING_ORIGINAL_EXPERIMENT`**<br></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fdcf256",
   "metadata": {},
   "source": [
    "Run **`SCRIPTS/3_DESEq2analysis_bootstrap_replicates_Jupyter.r`** in **`Rstudio`** or using  **`R`** in your terminal .\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The script creates:</b><br>\n",
    "\n",
    "- A folder called **`DESEq2_RESULTS`** within each of the folders inside **`PRE_PROCESSING_BS_REPLICATES`** <br>\n",
    "- Corresponding results from the **`DESEq2`** analyses (including **`LFC`** files) inside the **`DESEq2_RESULTS`** folder<br></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3c6b2d8",
   "metadata": {},
   "source": [
    "Run **`/SCRIPTS/3_DESEq2analysis_original_experiment_Jupyter.r`** in **`Rstudio`** or using **`R`** in your terminal.\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The script creates:</b><br>    \n",
    "    \n",
    "-  A folder called **`DESEq2_RESULTS`** within each of the folders inside **`PRE_PROCESSING_ORIGINAL_EXPERIMENT`** <br>\n",
    "- Corresponding results from the **`DESEq2`** analyses (including **`LFC`** files) inside the **`DESEq2_RESULTS`** folder<br></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92c5b847",
   "metadata": {},
   "source": [
    " <div class=\"alert alert-block alert-info\">\n",
    "<b>After running the last 4 scripts, the structure in the local computer at this point of the analysis is:</b>  \n",
    "    \n",
    "-  **`statSDS`**\n",
    " - **`ANALYSIS_RESULTS`** \n",
    "     - **`BOOTSTRAP_REPLICATES_OF_EXPERIMENT`**\n",
    "     - **`PRE_PROCESSING_BS_REPLICATES`**    \n",
    "         - **`DESEq2_RESULTS`**\n",
    "            - `lfc_Q111.txt`\n",
    "            - `lfc_Q170.txt`\n",
    "            - `lfc_Q175.txt`\n",
    "            - `lfc_Q50.txt`               \n",
    "            - `OTHER FILES`\n",
    "         - `count_dat.txt`\n",
    "         - `count_data.txt`\n",
    "         - `meta_dat.txt`\n",
    "     - **`PRE_PROCESSING_ORIGINAL_EXPERIMENT`**    \n",
    "         - (Same structure as **`PRE_PROCESSING_BS_REPLICATES`**)\n",
    " - **`EXAMPLE_INPUT_DATA`**\n",
    " - **`mousenet_V2_filtered`**\n",
    " -  **`statSDS_SCRIPTS`**     \n",
    " - **`SCRIPTS`** \n",
    " - **`launch_staSDS.ipynb`** \n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f550c8e",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94f9cf65",
   "metadata": {},
   "source": [
    "## 3 - Modify gene names if necessary, in local computer\n",
    "\n",
    "If your **`DESEq2`** output **`LFC`** files contain **`ENSEMBL IDs`** instead of **`Gene Symbols`**, and the reference bionetwork you want to use contains **`Gene Symbols`**, you need to convert **`ENSEMBL IDs`** to **`Gene Symbols`** so that you can launch **`Stat-SDS`**.  Skip this step otherwise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "477fdb96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " --------------------------------------------------------\n",
      "     CHANGING NAMES OF LFC FILES IN  BSreplicate_2\n",
      " --------------------------------------------------------\n",
      " --------------------------------------------------------\n",
      "     CHANGING NAMES OF LFC FILES IN  BSreplicate_1\n",
      " --------------------------------------------------------\n",
      " --------------------------------------------------------\n",
      "     CHANGING NAMES OF LFC FILES IN  BSreplicate_3\n",
      " --------------------------------------------------------\n",
      "\n",
      "\t              Script ended after 0:30:33.331053 \n",
      "      \n",
      "\t====================================================================         \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Run for bootstrap replicates' LFC files\n",
    "%run ./SCRIPTS/4_modify_gene_names_multiple_LFC_files.py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a05b644f",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The script creates:</b><br>   \n",
    "    \n",
    "LFC files ending with **`\"_geneSymbols.txt\"`** that contain **`Gene Symbol`**  nomenclature on the first column and corresponding **`LFC`** values on the second.<br></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "709d976c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\t              Script ended after 0:01:21.026471 \n",
      "  \n",
      "\t====================================================================         \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Run for original LFC file\n",
    "%run ./SCRIPTS/4_modify_gene_names_LFC_file.py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "227c644e",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The script creates:</b><br>    \n",
    "    \n",
    "LFC files ending with **`\"_geneSymbols.txt\"`** that contain **`Gene Symbol`**  nomenclature on the first column and corresponding **`LFC`** values on the second.<br></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31690d16",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b5db010",
   "metadata": {},
   "source": [
    "##  4 - Copy  `LFC` files from local computer to server\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "<b>Warning:</b> You will need to adapt the corresponding scripts so that they are match the structure of your computer and server, as well as the number of  `BS replicates ` you have created. </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8d49d3e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7382014b",
   "metadata": {},
   "source": [
    "First, we create directories in remote server to store **`LFC`** files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ec8c4ac8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> Creating folders in  remote server \n",
      "/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_original_experiment\n",
      "/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS1/\n",
      "/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS2/\n",
      "/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS3/\n"
     ]
    }
   ],
   "source": [
    "# Run script to create directories in remote server\n",
    "!./SCRIPTS/5_mkdir_several_to_server.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f63bb9d0",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>At this point of the analysis and considering the default example, the structure of the analysis folder in the server has to be:</b>  \n",
    "    \n",
    "-  **`StatSDS`**\n",
    " - **`mousenet_V2_filtered`**\n",
    "     - `geneNames.txt`\n",
    "     - `networkNames.txt`\n",
    " - **`fourier.py`**\n",
    " - **`genextract.py`**\n",
    " - **`launch_statSDS_V1.py`**\n",
    " - **`parameters.py`**\n",
    " - **`sds.py`**\n",
    " - **`sparse.py`**\n",
    " - **`statSDS_BS1`**\n",
    " - **`statSDS_BS2`**\n",
    " - **`statSDS_BS3`**\n",
    " - **`statSDS_original_experiment`**\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b1f3b07",
   "metadata": {},
   "source": [
    "Next, we copy **`LFC`** files (both those of the original experiment and of bootstrap replicates) to the folders we just created:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ae687cc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> Copying to server\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_ORIGINAL_EXPERIMENT/DESEq2_RESULTS/lfc_Q50_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_ORIGINAL_EXPERIMENT/DESEq2_RESULTS/lfc_Q111_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_ORIGINAL_EXPERIMENT/DESEq2_RESULTS/lfc_Q170_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_ORIGINAL_EXPERIMENT/DESEq2_RESULTS/lfc_Q175_geneSymbols.txt\n",
      "==> Copying to server\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_1/DESEq2_RESULTS/lfc_Q50_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_1/DESEq2_RESULTS/lfc_Q111_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_1/DESEq2_RESULTS/lfc_Q170_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_1/DESEq2_RESULTS/lfc_Q175_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_2/DESEq2_RESULTS/lfc_Q50_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_2/DESEq2_RESULTS/lfc_Q111_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_2/DESEq2_RESULTS/lfc_Q170_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_2/DESEq2_RESULTS/lfc_Q175_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_3/DESEq2_RESULTS/lfc_Q50_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_3/DESEq2_RESULTS/lfc_Q111_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_3/DESEq2_RESULTS/lfc_Q170_geneSymbols.txt\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/PRE_PROCESSING_BS_REPLICATES/BSreplicate_3/DESEq2_RESULTS/lfc_Q175_geneSymbols.txt\n"
     ]
    }
   ],
   "source": [
    "# You need to change the N parameter in the script to reflect the amount of BS replicates \n",
    "!./SCRIPTS/6_copy_several_to_server.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e0b22c1",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>At this point of the analysis and considering the default example, the structure of the analysis folder in the server has to be:</b>  \n",
    "    \n",
    "-  **`StatSDS`**\n",
    " - **`mousenet_V2_filtered`**\n",
    "     - `geneNames.txt`\n",
    "     - `networkNames.txt`\n",
    " - **`fourier.py`**\n",
    " - **`genextract.py`**\n",
    " - **`launch_statSDS_V1.py`**\n",
    " - **`parameters.py`**\n",
    " - **`sds.py`**\n",
    " - **`sparse.py`**\n",
    " - **`statSDS_BS1`**\n",
    "    - `Q111_LFC_geneSymbols.txt`\n",
    "    - `Q170_LFC_geneSymbols.txt`\n",
    "    - `Q175_LFC_geneSymbols.txt`\n",
    "    - `Q50_LFC_geneSymbols.txt`\n",
    " - **`statSDS_BS2`**\n",
    "    - `Q111_LFC_geneSymbols.txt`\n",
    "    - `Q170_LFC_geneSymbols.txt`\n",
    "    - `Q175_LFC_geneSymbols.txt`\n",
    "    - `Q50_LFC_geneSymbols.txt`\n",
    " - **`statSDS_BS3`**\n",
    "    - `Q111_LFC_geneSymbols.txt`\n",
    "    - `Q170_LFC_geneSymbols.txt`\n",
    "    - `Q175_LFC_geneSymbols.txt`\n",
    "    - `Q50_LFC_geneSymbols.txt`\n",
    " - **`statSDS_original_experiment`**\n",
    "    - `Q111_LFC_geneSymbols.txt`\n",
    "    - `Q170_LFC_geneSymbols.txt`\n",
    "    - `Q175_LFC_geneSymbols.txt`\n",
    "    - `Q50_LFC_geneSymbols.txt`\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cd9b2b6",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6b98da6",
   "metadata": {},
   "source": [
    "## 5 - Launch `Stat-SDS` on the server\n",
    "\n",
    "Copy script **`7_statSDS_run_SDS_original_data.sh`** to the **`StatSDS`** directory of the server.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>At this point of the analysis and considering the default example, the structure of the analysis folder in the server has to be:</b>  \n",
    "    \n",
    "-  **`StatSDS`**\n",
    " - **`mousenet_V2_filtered`**\n",
    "     - `geneNames.txt`\n",
    "     - `networkNames.txt`\n",
    " - **`fourier.py`**\n",
    " - **`genextract.py`**\n",
    " - **`launch_statSDS_V1.py`**\n",
    " - **`parameters.py`**\n",
    " - **`sds.py`**\n",
    " - **`sparse.py`**\n",
    " - **`statSDS_BS1`**\n",
    "    - `Q111_LFC_geneSymbols.txt`\n",
    "    - `Q170_LFC_geneSymbols.txt`\n",
    "    - `Q175_LFC_geneSymbols.txt`\n",
    "    - `Q50_LFC_geneSymbols.txt`\n",
    " - **`statSDS_BS2`**\n",
    "    - `Q111_LFC_geneSymbols.txt`\n",
    "    - `Q170_LFC_geneSymbols.txt`\n",
    "    - `Q175_LFC_geneSymbols.txt`\n",
    "    - `Q50_LFC_geneSymbols.txt`\n",
    " - **`statSDS_BS3`**\n",
    "    - `Q111_LFC_geneSymbols.txt`\n",
    "    - `Q170_LFC_geneSymbols.txt`\n",
    "    - `Q175_LFC_geneSymbols.txt`\n",
    "    - `Q50_LFC_geneSymbols.txt`\n",
    " - **`statSDS_original_experiment`**\n",
    "    - `Q111_LFC_geneSymbols.txt`\n",
    "    - `Q170_LFC_geneSymbols.txt`\n",
    "    - `Q175_LFC_geneSymbols.txt`\n",
    "    - `Q50_LFC_geneSymbols.txt`\n",
    "  - **`7_statSDS_run_SDS_original_data.sh`**  \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "facb6e86",
   "metadata": {},
   "source": [
    "Execute the script as a normal shell script with the following command: **`./7_statSDS_run_SDS_original_data.sh`**.\n",
    "\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>After running the script for the default configuration provided, the structure of the analysis folder in the server should be:</b>  \n",
    "    \n",
    "-  **`StatSDS`**\n",
    " - **`mousenet_V2_filtered`**\n",
    "     - `geneNames.txt`\n",
    "     - `networkNames.txt`\n",
    " - **`fourier.py`**\n",
    " - **`genextract.py`**\n",
    " - **`launch_statSDS_V1.py`**\n",
    " - **`parameters.py`**\n",
    " - **`sds.py`**\n",
    " - **`sparse.py`**\n",
    " - **`statSDS_BS1`**\n",
    "    - `Q111_LFC_geneSymbols.txt`\n",
    "    - `Q170_LFC_geneSymbols.txt`\n",
    "    - `Q175_LFC_geneSymbols.txt`\n",
    "    - `Q50_LFC_geneSymbols.txt`\n",
    "    - **`statSDS`**\n",
    "        - **`Q111_LFC_geneSymbols`**\n",
    "            - **`ABS`**\n",
    "                - `ABS_10.sif`\n",
    "                - `ABS_10.txt`   \n",
    "            - **`D`**\n",
    "                - `D_10.sif`\n",
    "                - `D_10.txt`\n",
    "            - **`U`**\n",
    "                - `U_10.sif`\n",
    "                - `U_10.txt`\n",
    "            - `expression.sparse`\n",
    "            - `graph.sparse`\n",
    "            - `means.txt`\n",
    "            - `smoothexpression.txt`\n",
    "            - `smoothexpression_10.txt`\n",
    "        - **`Q170_LFC_geneSymbols`**\n",
    "            - (Same structure as **`Q111_LFC_geneSymbols`**)\n",
    "        - **`Q175_LFC_geneSymbols`**\n",
    "            - (Same structure as **`Q111_LFC_geneSymbols`**)\n",
    "        - **`Q50_LFC_geneSymbols`**\n",
    "            - (Same structure as **`Q111_LFC_geneSymbols`**) \n",
    " - **`statSDS_BS2`**\n",
    "    - (Same structure as **`statSDS_BS1`**)\n",
    " - **`statSDS_BS3`**\n",
    "    - (Same structure as **`statSDS_BS1`**)\n",
    " - **`statSDS_original_experiment`**\n",
    "    - (Same structure as **`statSDS_BS1`**)\n",
    "  - **`7_statSDS_run_SDS_original_data.sh`**  \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f26ab4b",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f571b287",
   "metadata": {},
   "source": [
    "## 6 - Copy `Stat-SDS` results from server to local computer\n",
    "\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "<b>Warning:</b> You will need to adapt the corresponding scripts so that they are match the structure of your computer and server, as well as the number of  `BS replicates ` you have created. </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ada4d65b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> Copying from server\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_original_experiment/statSDS/Q50_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/original_experiment/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_original_experiment/statSDS/Q111_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/original_experiment/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_original_experiment/statSDS/Q170_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/original_experiment/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_original_experiment/statSDS/Q175_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/original_experiment/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS1/statSDS/Q50_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS1/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS1/statSDS/Q111_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS1/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS1/statSDS/Q170_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS1/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS1/statSDS/Q175_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS1/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS2/statSDS/Q50_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS2/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS2/statSDS/Q111_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS2/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS2/statSDS/Q170_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS2/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS2/statSDS/Q175_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS2/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS3/statSDS/Q50_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS3/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS3/statSDS/Q111_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS3/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS3/statSDS/Q170_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS3/\n",
      "satish@134.157.185.144:/home/satish/Desktop/MAIALEN/CODES/StatSDS/statSDS_BS3/statSDS/Q175_LFC_geneSymbols/\n",
      "/home/maialen/WorkSpace/CODING/statSDS_for_publication/ANALYSIS_RESULTS/SDS_NETWORKS/statSDS_BS3/\n"
     ]
    }
   ],
   "source": [
    "!./SCRIPTS/8_copy_several_from_server.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "773f7241",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "    \n",
    "<b>After copying **`Stat-SDS`** analysis results from server to local, the structure in the local computer at this point of the analysis is:</b>  \n",
    "    \n",
    "-  **`statSDS`**\n",
    " - **`ANALYSIS_RESULTS`** \n",
    "     - **`BOOTSTRAP_REPLICATES_OF_EXPERIMENT`**\n",
    "     - **`PRE_PROCESSING_BS_REPLICATES`**\n",
    "     - **`PRE_PROCESSING_ORIGINAL_EXPERIMENT`**\n",
    "     - **`SDS_NETWORKS`**\n",
    "        - **`original_experiment`**\n",
    "            - **`Q50_LFC_geneSymbols`**\n",
    "                - **`ABS`**\n",
    "                    - `ABS_10.sif`\n",
    "                    - `ABS_10.txt`   \n",
    "                - **`D`**\n",
    "                    - `D_10.sif`\n",
    "                    - `D_10.txt`\n",
    "                - **`U`**\n",
    "                    - `U_10.sif`\n",
    "                    - `U_10.txt`\n",
    "                - `expression.sparse`\n",
    "                - `graph.sparse`\n",
    "                - `means.txt`\n",
    "                - `smoothexpression.txt`\n",
    "                - `smoothexpression_10.txt`\n",
    "            - **`Q111_LFC_geneSymbols`**\n",
    "                - (Same structure as **`Q50_LFC_geneSymbols`**)\n",
    "            - **`Q170_LFC_geneSymbols`**\n",
    "                - (Same structure as**` Q50_LFC_geneSymbols`**)\n",
    "            - **`Q175_LFC_geneSymbols`**\n",
    "                - (Same structure as **`Q50_LFC_geneSymbols`**)\n",
    "    \n",
    "        - **`statSDS_BS1`**\n",
    "             - (Same structure as **`original_experiment`**)\n",
    "\n",
    "        - **`statSDS_BS2`**\n",
    "             - (Same structure as **`original_experiment`**)\n",
    "        - **`statSDS_BS3`**\n",
    "             - (Same structure as **`original_experiment`**)\n",
    "    \n",
    " - **`EXAMPLE_INPUT_DATA`**\n",
    " - **`mousenet_V2_filtered`**\n",
    " -  **`statSDS_SCRIPTS`**     \n",
    " - **`SCRIPTS`** \n",
    " - **`launch_staSDS.ipynb`** \n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66bb43a5",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a24a6432",
   "metadata": {},
   "source": [
    "## 7 - Count number of BS replicate `Stat-SDS` networks containing a given edge\n",
    "\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "<b>Warning:</b> You have to parse the number of `BS replicates` you selected on Step 1 of the workflow under the **`--n_bs_replicates`** argument when running the script as well as the type of deregulaton for which you want to perform the filtering of edges under the argument **`--deregulation`**. </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2d04535b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of bootstrap replicates generated for the experiment = 3\n",
      "Type of deregulated network to analyse = ABS\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 1 of Q50\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 2 of Q50\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 3 of Q50\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 1 of Q111\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 2 of Q111\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 3 of Q111\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 1 of Q170\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 2 of Q170\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 3 of Q170\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 1 of Q175\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 2 of Q175\n",
      " -------------------------------------------\n",
      "  Analysing BS network replicate 3 of Q175\n",
      "\n",
      "\t==================================================================== \n",
      "\n",
      "\t              Script ended after 0:00:34.897619 \n",
      "      \n",
      "\t====================================================================         \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%run ./SCRIPTS/9_calculate_BSreplicates_edge_presence_Jupyter.py --n_bs_replicates 3 --deregulation ABS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e51502f4",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The script creates:</b> \n",
    "    \n",
    "A folder named **`BS_REPLICATES_EDGE_PRESENCE`** within the **`ANALYSIS_RESULTS`** containing SIF and CSV files representing the number of **`BS replicate**` networks in which a given edge is found. \n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "    \n",
    "<b>The structure of the **`statSDS`** analysis folder in your local computer after running the script should be:</b>  \n",
    "    \n",
    "-  **`statSDS`**\n",
    " - **`ANALYSIS_RESULTS`** \n",
    "     - **`BOOTSTRAP_REPLICATES_OF_EXPERIMENT`**\n",
    "     - **`PRE_PROCESSING_BS_REPLICATES`**\n",
    "     - **`PRE_PROCESSING_ORIGINAL_EXPERIMENT`**\n",
    "     - **`SDS_NETWORKS`**\n",
    "     - **`BS_REPLICATES_EDGE_PRESENCE`**\n",
    "        - `BSreplicates_edge_presence_scores_Q50ABS` `SIF` and `CSV` files\n",
    "        - `BSreplicates_edge_presence_scores_Q111ABS` `SIF` and `CSV` files\n",
    "        - `BSreplicates_edge_presence_scores_Q170ABS` `SIF` and `CSV` files\n",
    "        - `BSreplicates_edge_presence_scores_Q175ABS` `SIF` and `CSV` files\n",
    "        \n",
    "    \n",
    " - **`EXAMPLE_INPUT_DATA`**\n",
    " - **`mousenet_V2_filtered`**\n",
    " -  **`statSDS_SCRIPTS`**     \n",
    " - **`SCRIPTS`** \n",
    " - **`launch_staSDS.ipynb`** \n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "743a1368",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84abd62c",
   "metadata": {},
   "source": [
    "## 8 - Filter `Stat-SDS` network edges given a certain threshold for edge presence\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "<b>Warning:</b> You have to parse the edge presence threshold (any number between `1` and `100` representing the percentage of bootstrap replicates that contain a given edge) under the **`--e`** argument when running the script as well as the type of deregulaton for which you want to perform the filtering of edges under the argument **`--deregulation`**. </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b32ec09e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge presence threshold for network filtering:  80.0\n",
      "Deregulated networks to be filtered:  ABS\n",
      " -----------------------------------\n",
      "       Filtering Q50 ABS network\n",
      " -----------------------------------\n",
      "       Filtering Q111 ABS network\n",
      " -----------------------------------\n",
      "       Filtering Q170 ABS network\n",
      " -----------------------------------\n",
      "       Filtering Q175 ABS network\n",
      "\n",
      "\t==================================================================== \n",
      "\n",
      "\t              Script ended after 0:08:48.706528 \n",
      "      \n",
      "\t====================================================================         \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%run ./SCRIPTS/10_filter_network_BSreplicates_reference_percentage_Jupyter.py -e 80 -deregulation ABS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d3fd496",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>The script creates:</b> \n",
    "    \n",
    "A folder named **`FILTERED_NETWORKS`** within the **`ANALYSIS_RESULTS`** where networks filtered for the edgge presence threshold are stored. \n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "    \n",
    "<b>After filtering the edges of the networks based on the chosen threshold, the structure in the local computer at this point of the analysis is:</b>  \n",
    "    \n",
    "-  **`statSDS`**\n",
    " - **`ANALYSIS_RESULTS`** \n",
    "     - **`BOOTSTRAP_REPLICATES_OF_EXPERIMENT`**\n",
    "     - **`PRE_PROCESSING_BS_REPLICATES`**\n",
    "     - **`PRE_PROCESSING_ORIGINAL_EXPERIMENT`**\n",
    "     - **`SDS_NETWORKS`**\n",
    "     - **`BS_REPLICATES_EDGE_PRESENCE`**\n",
    "     - **`FILTERED_NETWORKS`**    \n",
    "        - `filtered_network_80.0percentageBSreps_Q50ABS.sif` \n",
    "        - `filtered_network_80.0percentageBSreps_Q111ABS.sif`    \n",
    "        - `filtered_network_80.0percentageBSreps_Q17ABS.sif`    \n",
    "        - `filtered_network_80.0percentageBSreps_Q175ABS.sif`        \n",
    " - **`EXAMPLE_INPUT_DATA`**\n",
    " - **`mousenet_V2_filtered`**\n",
    " -  **`statSDS_SCRIPTS`**     \n",
    " - **`SCRIPTS`** \n",
    " - **`launch_staSDS.ipynb`** \n",
    "    \n",
    "</div>\n",
    "    </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f1c7370",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0d52d64",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Congratulatioms:</b> \n",
    "You should now have filtered networks within the  `ANALYSIS_RESULTS/FILTERED_NETWORKS ` folder that you can use for further data analyses.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9e3f72a",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e1bc1ca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
